File: src/main/java/com/mercateo/jsonschema/generictype/GenericClass.java
package com.mercateo.jsonschema.generictype;

import com.googlecode.gentyref.GenericTypeReflector;

final class GenericClass<T> extends GenericTypeAbstract<T, Class<?>> {

    GenericClass(Class<T> type) {
        super(type, type);
    }

    @Override
    public String getSimpleName() {
        return getRawType().getSimpleName();
    }

    @Override
    public GenericType<?> getContainedType() {
        if (getRawType().isArray()) {
            return GenericType.of(GenericTypeReflector.getArrayComponentType(getRawType()),
                    getRawType().getComponentType());
        }
        throw new IllegalAccessError("GenericClass " + getSimpleName() + " has no contained type");
    }

    @Override
    public GenericType<? super T> getSuperType() {
        final Class<? super T> superclass = getRawType().getSuperclass();
        return superclass != null ? GenericType.of(GenericTypeReflector.getExactSuperType(
                getRawType(), superclass), superclass) : null;
    }

    @Override
    public boolean isIterable() {
        return getRawType().isArray();
    }

    @Override
    public String toString() {
        return getRawType().toString();
    }
}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericClass.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector

internal class GenericClass<T>(type: Class<T>) : GenericTypeAbstract<T, Class<*>>(type, type) {

    override val containedType: GenericType<*>
        get() {
            if (rawType.isArray) {
                return GenericType.of(
                    GenericTypeReflector.getArrayComponentType(rawType),
                    rawType.componentType
                )
            }
            throw IllegalAccessError("GenericClass $simpleName has no contained type")
        }

    override val isIterable: Boolean
        get() = rawType.isArray

    override fun toString(): String {
        return rawType.toString()
    }
}
